txpool/README.md

```diff
@@ -0,0 +1,60 @@
+* by Dependency: With every fuel transaction, inputs and outputs change state and we need to be aware of it. Graph is main defensive structure behind ddos, every transaction that we include in pool should have potential to be included in next block in that sense Graph represent connection between parent and child where child is transaction that depends on execution output of parent that is found inside database or transaction pool.
+
+## Dependency graph
+
+Few reasonings on decision and restrains made by txpool
+
+Visualization of dependencies between three transactions:
+![UTXO dependency diagram](../../docs/diagrams/fuel_v2_client_design_UTXO_dependency.jpg)
+
+Problem1: T2 arriving before T1.
+Solution: Have restriction that broadcast of hashes need to contains ancestor hashes and they all need to be sorted by gasPby_hashrice
+
+Problem2: if T2 gas price is higher then T1. This can be problematic on txpool with different sizes. One big enough pull could contains T1 but other one would prune it and we could not prove that T2 can be included without T1.
+Solution: Have restriction on linked transaction within same block so that GasPrice is strictly going down. This will effect newly create contract (created in same block) but not old one.
+
+Usage: Insertion of new T4
+
+* Usage: inserting T4 that replaces T1. It is done if GasPrice(T4) > GasPrice(T1) and if there is some overlaping inputs between them AnyInput(T4) == AnyInput(T1).
+* Usage: inserting T4 that use State6 and GasPrice(T4) > GasPrice(T3).
+    Result: Remove T3 from txpool check every T3 child for removal. Check if any T that has lover gas is reusing T4 outputs so thay should be connected to T4.
+    
+### Implementation
+
+Graph will be needed for relationship between dependable transactions so that if one of them get included/removed/replaced we need to know what to do with their dependent childrens.
+We will need map of all outputs in same structure so that we can fast search them. That would be something like `HashMap<UtxoId, CoinState>` and `HashMap<ContractId,ContractState>` for both Coin and Contract state.
+
+### Block inclusion Algorithm:
+It is most straightforward one: take PriceSort and iterate over transaction. DependencyGraph inclusion garantee us that every transaction in that sorted array can be included, but only after execution that trasansaction we would be sure that is can go inside block.
+
+### refs:
+* Some information on OpenEthereum and restrictions on tx inclusion: https://openethereum.github.io/Transactions-Queue
+* Possible out of scope, but it is relatively easy to add ethereum's eip1559 that adds Fee market change with basefee. More on eip1559: https://hackmd.io/@timbeiko/1559-resources
+Erigon did good docs on impl of eip1559 txpool: https://github.com/ledgerwatch/erigon/wiki/Transaction-Pool-Design
```

Comment:
Should we have a section about future work that may be required? Like managing pending txs that aren't currently valid and waiting for other data, or byte prices independent of gas price?

Reply:
Good idea, will be added